### ISSUE TITLE
Audio glitch during menu fade in Chrono Trigger

### ISSUE THREAD

**Issue by realnc** [snes9xgit/snes9x#755]
When pressing X in Chrono Trigger to activate the menu, and B to go out of the menu, during the fade in/out of the menu, audio glitches occur. This is with snes9x-gtk in Gentoo Linux using any audio driver. It also occurs when using the libretro core.

A quick look shows that too many samples per frame are being generated during the menu fade, followed a couple frames later by too few samples.

I tested latest git master as well as 1.61.

How to reproduce:

Start Chrono Trigger (USA). Once you get control of the main character, press X to show the menu, and then B to exit the menu. During the menu fade, audio glitches should occur.

**Comment by bearoso** [comment]
Interesting. I'm going to guess that the emulator doesn't drop out to the frontend during one of those black frames and it runs too long. That'll throw the sound timing off. I'd bet if the screen wasn't black it would glitch there, too.

Some operation is taking a long time. I'll have to see what it is. This bug's probably been there forever.

**Comment by realnc** [comment]
Note that the video I posted is not from the libretro core. It's snes9x-gtk.

**Comment by bearoso** [comment]
The problem here is that there's a long DMA during the fade-outs. S9xMainLoop doesn't return. This affects the libretro core because it no longer uses the sample callback, so more than one frame's sound data accumulates. I changed it to instead process the sound data during the video frame callback at the end, which produces a consistent 1066/1068 samples now.

As for snes9x-gtk, it *shouldn't* produce the glitch. It drains the samples every scanline with the SamplesAvailable callback, which isn't reliant on S9xMainLoop returning. I am unable to reproduce the behavior seen in your video. Can you tell me the driver and settings you're using for that?

**Comment by realnc** [comment]
> As for snes9x-gtk, it _shouldn't_ produce the glitch. It drains the samples every scanline with the SamplesAvailable callback, which isn't reliant on S9xMainLoop returning. I am unable to reproduce the behavior seen in your video. Can you tell me the driver and settings you're using for that?

Default settings (`rm ~/.config/snes9x/snes9x.conf`) with PulseAudio or SDL audio driver. The audio is generally very "crackly," but it helps to raise the audio buffer size to 64ms to reduce it. The audio dropout during the menu fade remains though.

It sounded similar to the libretro core issue so it appeared to be the same cause. But apparently that's not true.

**Comment by realnc** [comment]
I tried your latest commit and it fixes the issue, but it breaks audio when runahead is enabled. There's no audio at all with second instance runahead, and FPS is halved without second instance.

**Comment by realnc** [comment]
I submitted a PR to fix the issue: https://github.com/snes9xgit/snes9x/pull/760

**Comment by bearoso** [comment]
That looks OK. The real fix would be to rework the DMA code to be a state machine that can we can break out of when necessary. That way S9xMainLoop is always exactly one frame. Or use threads or cothreads, but that would be platform-specific or necessitate the upgrade of the core to C++11 for std::thread.

**Comment by realnc** [comment]
The oldest platform the core is offered for is GameCube (devkitpro) which does support C++17. Older platforms than that use the C-based forks of Snes9x. So it should be fine to upgrade to modern C++, although I haven't tested std::thread support in devkitpro.

The docker images that are used for building are:

GameCube and Wii: git.libretro.com:5050/libretro-infrastructure/libretro-build-devkitpro:latest
WiiU: git.libretro.com:5050/libretro-infrastructure/libretro-build-wiiu:latest
Switch: git.libretro.com:5050/libretro-infrastructure/libretro-build-libnx-devkitpro:latest

**Comment by realnc** [comment]
A quick build of the core for GameCube with this simple test builds fine:

```diff
diff --git a/libretro/Makefile b/libretro/Makefile
index 69e8899..002a1fc 100644
--- a/libretro/Makefile
+++ b/libretro/Makefile
@@ -612,7 +612,8 @@ CXXFLAGS	+= $(CODE_DEFINES) $(WARNINGS_DEFINES) $(fpic)
 CXXFLAGS	+= -DRIGHTSHIFT_IS_SAR -D__LIBRETRO__ -DALLOW_CPU_OVERCLOCK
 CFLAGS		:= $(CXXFLAGS)
 CFLAGS          += -DHAVE_STDINT_H
-CXXFLAGS        += -DHAVE_STDINT_H
+CXXFLAGS        += -DHAVE_STDINT_H -std=c++17
+LDFLAGS         += -Wl,--no-undefined
 ifeq (,$(findstring msvc,$(platform)))
 ifeq ($(HAVE_STRINGS_H), 1)
 CXXFLAGS += -DHAVE_STRINGS_H
diff --git a/libretro/libretro.cpp b/libretro/libretro.cpp
index 5802841..e5c5512 100644
--- a/libretro/libretro.cpp
+++ b/libretro/libretro.cpp
@@ -82,8 +82,12 @@ const int MAX_SNES_WIDTH_NTSC = ((SNES_NTSC_OUT_WIDTH(256) + 3) / 4) * 4;
 static bool show_lightgun_settings = true;
 static bool show_advanced_av_settings = true;
 
+#include <thread>
 static void extract_basename(char *buf, const char *path, size_t size)
 {
+    std::thread t([]{ });
+    t.join();
+
     const char *base = strrchr(path, '/');
     if (!base)
         base = strrchr(path, '\\');
```

**Comment by realnc** [comment]
Alright, so it turns out the problem goes away if I use opengl instead of software (software was the default previously), then switch my monitor to 60Hz (I use 120Hz normally), then enable opengl vsync and change "throttling method" from "timer based" to "no throttling, use vsync".

**Comment by bearoso** [comment]
This is sort of fixed now. From at least as far back as the first commit in git, S9xSyncSpeed, which was responsible for controlling the speed, was being called only as S9xMainLoop exited. So this issue where the DMA lasts more than one frame was affecting anything that used it to throttle, running three frames at once and causing the sound buffer to overflow. The problem goes away if you use opengl because vsync takes place when Snes9x calls S9xDeinitUpdate every frame. 

So I've removed the extraneous callback and used S9xSyncSpeed to handle audio in the libretro port. Problem solved.

**Comment by OV2** [comment]
Relevant commits: 50f9f9eba12586bad3c4d029ffb3363c48bed5ac and 16a45f7b92efdb88c4d3ebee784e446118e8c4ad

Don't we have a similar issue with input during those frames (at least for buttons not mapped with poll=true) ? Since we only call S9xReportButton outside of it?

**Comment by bearoso** [comment]
I don't think so. If a DMA is running multiple frames, the game is most likely not processing input, and definitely doesn't need updated control info at that point. Remember, it breaks out of the main loop right after the DMA because the SCAN_KEYS flag is still set, even if it's already halfway into a frame.
